### DISCUSSION TITLE
Install HSSM in Mac with M1 chip

### DISCUSSION THREAD

**Discussion by Hellobamboobamboo** [lnccbrown/HSSM#495] [General]
Which code is for installing HSSM in Mac with M1 chip, the GPU one? The code on the directed website doesn't seem to work...

**Comment by Hellobamboobamboo** [comment]
hannels:
 - conda-forge
 - defaults
Platform: osx-arm64
Collecting package metadata (repodata.json): done
Solving environment: \ warning  libmamba Added empty dependency for problem type SOLVER_RULE_UPDATE
failed

LibMambaUnsatisfiableError: Encountered problems while solving:
  - nothing provides _python_rc needed by python-3.12.0rc3-rc3_h47c9636_1_cpython

Could not solve for environment specs
The following packages are incompatible
├─ hssm is installable and it requires
│  └─ python >=3.10,<3.12  with the potential options
│     ├─ python [3.10.0|3.10.1|...|3.11.9], which can be installed;
│     └─ python 3.12.0rc3 would require
│        └─ _python_rc, which does not exist (perhaps a missing channel);
└─ pin-1 is not installable because it requires
   └─ python 3.12.* , which conflicts with any installable versions previously reported.

(base) jingzhuchen@vpn1722514025 ~ %

**Comment by Hellobamboobamboo** [comment]
Seems like the problem lies in "_python_rc, which does not exist"?

**Comment by Hellobamboobamboo** [comment]
Oh I solved it by downloading supporting files, and changed python to NOT the latest version. Whoop!

**Reply by hcp4715** [reply]
hi, @Hellobamboobamboo 
Thanks for sharing your experience. How is the speed of sample? It seems that the simplest model in the tutorial notebook need more than 2 hours. very weird.

Chuan-Peng

**Reply by digicosmos86** [reply]
Hi @hcp4715,

That is indeed strange. Can you provide the following information:

1. How did you install hssm? The only official way is `conda install -c conda-forge hssm`. `pip install hssm` is likely going to give you issues mentioned below:
2. When you run `import hssm`, did you see this warning? `WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.`

Thanks!
Paul

**Reply by hcp4715** [reply]
Hi, Paul,

thanks for your prompt response! Here is the info of my machine and system:
OS: macOS Sonoma version 14.6.1
Chip: Apple M1

I tried two ways to install hssm, please see below. Your suggestions are highly appreciated.

(1) using `conda` 

```
conda create -n hssm python=3.11
conda activate hssm

conda install -c conda-forge hssm
```
Unfortunately, I encountered an error when the following code of the main tutorial 
```
simple_ddm_model = hssm.HSSM(data=dataset)
```

Error message:
```
/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/hssm/distribution_utils/dist.py:482: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
  data_vector = pt.dvector()
/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pymc/distributions/continuous.py:309: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
  pt.bitwise_and(pt.ge(value, lower), pt.le(value, upper)),
/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/raise_op.py:97: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
  [value.type()],
/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/compile/function/pfunc.py:190: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
  clone_node_and_cache(
/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/compile/function/types.py:1414: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
  rewriter_profile = rewriter(fgraph)
/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/compile/function/types.py:1414: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
  rewriter_profile = rewriter(fgraph)
/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/compile/function/types.py:1414: UserWarning: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
  rewriter_profile = rewriter(fgraph)
ERROR (pytensor.graph.rewriting.basic): Rewrite failure due to: constant_folding
ERROR (pytensor.graph.rewriting.basic): node: Le(0.0, 20.0)
ERROR (pytensor.graph.rewriting.basic): TRACEBACK:
ERROR (pytensor.graph.rewriting.basic): Traceback (most recent call last):
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/graph/rewriting/basic.py", line 1913, in process_node
    replacements = node_rewriter.transform(fgraph, node)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/graph/rewriting/basic.py", line 1085, in transform
    return self.fn(fgraph, node)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/tensor/rewriting/basic.py", line 1166, in constant_folding
    return unconditional_constant_folding.transform(fgraph, node)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/graph/rewriting/basic.py", line 1085, in transform
    return self.fn(fgraph, node)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/tensor/rewriting/basic.py", line 1115, in unconditional_constant_folding
    thunk = node.op.make_thunk(node, storage_map, compute_map, no_recycling=[])
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/op.py", line 119, in make_thunk
    return self.make_c_thunk(node, storage_map, compute_map, no_recycling)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/op.py", line 84, in make_c_thunk
    outputs = cl.make_thunk(
              ^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/basic.py", line 1186, in make_thunk
    cthunk, module, in_storage, out_storage, error_storage = self.__compile__(
                                                             ^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/basic.py", line 1103, in __compile__
    thunk, module = self.cthunk_factory(
                    ^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/basic.py", line 1631, in cthunk_factory
    module = cache.module_from_key(key=key, lnk=self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/cmodule.py", line 1255, in module_from_key
    module = lnk.compile_cmodule(location)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/basic.py", line 1532, in compile_cmodule
    module = c_compiler.compile_str(
             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/cmodule.py", line 2685, in compile_str
    raise CompileError(
pytensor.link.c.exceptions.CompileError: Compilation failed (return status=1):
/usr/bin/clang++ -dynamiclib -g -O3 -fno-math-errno -Wno-unused-label -Wno-unused-variable -Wno-write-strings -Wno-c++11-narrowing -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -fPIC -undefined dynamic_lookup -I/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/numpy/core/include -I/Users/hcp4715/miniconda3/envs/hssm/include/python3.11 -I/Users/hcp4715/miniconda3/envs/hssm/lib/python3.11/site-packages/pytensor/link/c/c_code -L/Users/hcp4715/miniconda3/envs/hssm/lib -fvisibility=hidden -o /Users/hcp4715/.pytensor/compiledir_macOS-14.6.1-arm64-arm-64bit-arm-3.11.11-64/tmpue45o932/m8bf87d9c1073341b558a9998c11310cc1d482c7c54e1ae0c2d6a6616b80e05a3.so /Users/hcp4715/.pytensor/compiledir_macOS-14.6.1-arm64-arm-64bit-arm-3.11.11-64/tmpue45o932/mod.cpp
/Users/hcp4715/.pytensor/compiledir_macOS-14.6.1-arm64-arm-64bit-arm-3.11.11-64/tmpue45o932/mod.cpp:2:10: fatal error: 'iostream' file not found
    2 | #include <iostream>
      |          ^~~~~~~~~~
1 error generated.

...
You can find the C code in this temporary file: /var/folders/1_/hd9cg0112c9g8jqzmm8pjrqh0000gn/T/pytensor_compilation_error_xgowaj1y

You can find the C code in this temporary file: /var/folders/1_/hd9cg0112c9g8jqzmm8pjrqh0000gn/T/pytensor_compilation_error_2bbk2be1

You can find the C code in this temporary file: /var/folders/1_/hd9cg0112c9g8jqzmm8pjrqh0000gn/T/pytensor_compilation_error_ex6g4is1

You can find the C code in this temporary file: /var/folders/1_/hd9cg0112c9g8jqzmm8pjrqh0000gn/T/pytensor_compilation_error_uge78cgh
Model initialized successfully.
```
using `conda list` showed that I have install pytensor:

```
pytensor                  2.26.4          py311h8f8510f_0    conda-forge
pytensor-base             2.26.4          py311h4b98486_0    conda-forge
```


(2) I switched docker, used `quay.io/jupyter/minimal-notebook:aarch64-python-3.11.10` as based image and use `pip` to install `hssm`.  Indeed, when `import hssm` and , there were warnings:

```
WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.
[/opt/conda/lib/python3.11/site-packages/tqdm/auto.py:21](http://127.0.0.1:8888/opt/conda/lib/python3.11/site-packages/tqdm/auto.py#line=20): TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/opt/conda/lib/python3.11/site-packages/numpyro/ops/provenance.py:99: DeprecationWarning: jax.core.call_p is deprecated. Use jax.extend.core.primitives.call_p
  track_deps_rules[core.call_p] = track_deps_call_rule
[/opt/conda/lib/python3.11/site-packages/numpyro/ops/provenance.py:107](http://127.0.0.1:8888/opt/conda/lib/python3.11/site-packages/numpyro/ops/provenance.py#line=106): DeprecationWarning: jax.core.closed_call_p is deprecated. Use jax.extend.core.primitives.closed_call_p
  track_deps_rules[core.closed_call_p] = track_deps_closed_call_rule
```

PS: I tried to use another base image, `quay.io/jupyter/scipy-notebook:aarch64-python-3.11.10`, and then got different warnings, probably because the version of these modules? But the sample speed is basically the same.

<img width="1059" alt="image" src="https://github.com/user-attachments/assets/9f646498-4787-434b-a098-d6765ea7eeff" />


Best
Chuan-Peng

**Reply by AlexanderFengler** [reply]
@digicosmos86 do we have any idea what's going on there?

My hunch is that this is a `pymc` / `jax` problem, with lagging M1 GPU support, but not fully up to date on those developments.

**Reply by digicosmos86** [reply]
Hi @hcp4715,

I tried to start a new conda environment and installed hssm since I am also using an M1 machine. However, I wasn't able to replicate the error that you experienced. I noticed in your error that it is pointing to some C/C++ code compilation, which made me think that the error could be caused by the C/C++ compiler on your system. Here's a few things that I would try:

1. Have you installed xcode? If not, doing so could install everything required to compile C/C++ code.
2. If the above doesn't fix the problem, I would try a different container that only supports conda, like the [official anaconda image](https://docs.anaconda.com/working-with-conda/applications/docker/). They have a tutorial that shows how to get Jupyter installed. You only need to add HSSM to the mix and you won't have any issues that comes with using `pip install`

Please let me know if any of this works.

Thanks!
Paul

**Reply by hcp4715** [reply]
Hi, @digicosmos86, @AlexanderFengler 

Thanks a lot for both of you! 

> Have you installed xcode? If not, doing so could install everything required to compile C/C++ code.

I've installed Xcode but the error persist.

>If the above doesn't fix the problem, I would try a different container that only supports conda, like the [official anaconda image](https://docs.anaconda.com/working-with-conda/applications/docker/). They have a tutorial that shows how to get Jupyter installed. You only need to add HSSM to the mix and you won't have any issues that comes with using `pip install`

I also tried the official docker image of anaconda, but used the [image with python 11](https://hub.docker.com/layers/continuumio/miniconda3/23.10.0-1/images/sha256-166ff37fba6c25fcad8516aa5481a2a8dfde11370f81b245c1e2e8002e68bcce). The operating system of anaconda's image is also Linux and I have similar warnings as before:

<img width="1171" alt="image" src="https://github.com/user-attachments/assets/dbaa9acd-b166-4291-969f-7bb4b4a791d7" />

The speed of sampling is similar to my previous test too:
<img width="886" alt="image" src="https://github.com/user-attachments/assets/7d858a2f-13f3-47b6-be63-6d25bc6714ed" />

**Reply by digicosmos86** [reply]
> I also tried the official docker image of anaconda, but used the [image with python 11](https://hub.docker.com/layers/continuumio/miniconda3/23.10.0-1/images/sha256-166ff37fba6c25fcad8516aa5481a2a8dfde11370f81b245c1e2e8002e68bcce). The operating system of anaconda's image is also Linux and I have similar warnings as before:

Hi @hcp4715,

You need to use `conda install -c conda-forge hssm` in the anaconda container. We don't recommend installing hssm with `pip` under most circumstances, including within containers.

**Reply by hcp4715** [reply]
Hi, @digicosmos86, 

Thanks for your prompt response. `conda install -c conda-forge hssm` does not work, I guess it was because `conda-forge` does not have the package for platform `linux-aarch64`. 

<img width="689" alt="image" src="https://github.com/user-attachments/assets/24ebef50-72b6-42a5-88fa-fb7eef6f510a" />


But it works for using conda directly in MacOS.
<img width="543" alt="image" src="https://github.com/user-attachments/assets/e94db351-3536-4d05-83b1-089d7f41b81f" />

**Reply by digicosmos86** [reply]
Okay let's try this workaround:

First, you can install PyMC using conda: `conda install -c conda-forge pymc`
Then you can install hssm using pip into the same conda environment: `pip install hssm`.

The issue is basically getting all the linear algebra packages to work with PyMC, for which you can use conda. Then installing hssm with pip gets around the issue with one dependency that's not built for linux/aarch64 in conda.

You can try this locally first and see if a pip-installed HSSM fixes the issues. If not, then feel free to try it in the container.

Let me know if this works.

Thanks!
Paul

**Reply by hcp4715** [reply]
Hi, @digicosmos86 ,

I tried to install `pymc` using `conda install -c conda-forge pymc`, unfortunately it did not solve the issue. 

When I run it locally, still have this error:

![image](https://github.com/user-attachments/assets/85b1ca39-13a2-459f-aaf5-77433904b676)

When tried it in the conda's docker image, it was still slow:
![image](https://github.com/user-attachments/assets/c0419a3f-afba-45ce-9dbd-2ad05619d0f6)

As you can see from screenshot below, pymc was install from conda-forge:
![image](https://github.com/user-attachments/assets/28bb006c-ebb7-4e06-9967-3dfcc0c0cd4c)

**Reply by digicosmos86** [reply]
Okay the screen shots are helpful. It seems that the warnings are gone, so at least we solved that issue. Have you tried using the `nuts_numpyro` sampler and see if that results in any speed-up?

**Reply by hcp4715** [reply]
Hi, @digicosmos86 ,

Since you mentioned that the issue might be related to`pymc`, so I check pymc forum and indeed, the error I encountered [here](https://github.com/lnccbrown/HSSM/discussions/495#discussioncomment-11817146) was discussed there: https://discourse.pymc.io/t/problems-with-installation-of-pymc-after-xcode-update-on-mac/15097/3.

I re-installed Xcode and added `pytensor.config.cxx = "/usr/bin/clang++"`. now it finally worked, but it was still slow:

<img width="1183" alt="image" src="https://github.com/user-attachments/assets/815496cd-fe4e-4cda-aaef-57b4d9499cfb" />

so, I tried `numpyro`, and it is indeed much faster.

<img width="809" alt="image" src="https://github.com/user-attachments/assets/e6499fce-b26b-4c26-a3e4-f23f4fb33f89" />
